Add CLI support to run migrations.

vedanshujain opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and why such changes. -->

Add support to run migration via the command line. Since the backfill is not implemented, only migration from wp_posts and wp_postmeta to COT is currently supported.

Closes #32917 

### How to test the changes in this Pull Request:

1. Check out this branch and run `composer dump-autoload`. 
2. Run command `wp wc cot count_unmigrated`. You should see an output like `There are 22872 orders to be migrated.`.
3. Run the command `wp wc cot migrate` . You should see an output like 
```
This feature is not production ready yet. Make sure you are not running these commands in your production environment.
There are 22872 orders to be migrated.
Order Data Migration  100% [===========================================================================================================] 1:23 / 1:22
Migration completed.
Success: 22872 orders were migrated in 80.901494 seconds
```
4. Run the command `wp wc cot verify_cot_data`. You should an output like:
```
This feature is not production ready yet. Make sure you are not running these commands in your production environment.
There are 22872 orders to be verified.
Order Data Verification  100% [==============================================================================================================================] 1:39 / 1:38
Verification completed.
Success: 22872 orders were verified in 100.426563 seconds
```
5. Change some random order column value in any of the migrated table and run the verification command again using `wp wc cot verify_cot_data`. It should be able to pick up the modified data like:
```
This feature is not production ready yet. Make sure you are not running these commands in your production environment.
There are 22872 orders to be verified.
Order Data Verification  100% [==============================================================================================================================] 1:50 / 1:31
Verification completed.
Error: 22872 orders were verified in 110.967540 seconds. Errors were found Array
(
    [21145] => Array
        (
            [0] => _order_shipping_tax
        )

)
```
### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?
* [x] Have you created a changelog file by running `pnpm nx affected --target=changelog`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

vedanshujain committed:
Add CLI support to run migrations.

vedanshujain committed:
Add changelog.

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,241 @@
+	/**
+	 * Helper method to log warning that feature is not yet production ready.
+	 */
+	private function log_product_warning() {
```
It's quite confusing to shorten `production` to `product` in the context of an e-commerce software, maybe just use the full name?

Would it make sense to add a check for `wp_get_environment_type` here as well?

vedanshujain replied:
Addressed in https://github.com/woocommerce/woocommerce/pull/32938/commits/124fdd37f35389dd7ac8ba92255bb159dbc26b9b. Although didn't added a check for `wp_get_environment_type` as it felt like it would be one more hurdle in getting test results, wdyt?

peterfabian replied:
Yup, that sounds reasonable. 👍

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,241 @@
+			);
+			$batch_start_time = microtime( true );
+			$order_ids        = $this->synchronizer->get_ids_of_orders_pending_sync( $this->synchronizer::ID_TYPE_MISSING_IN_ORDERS_TABLE, $batch_size );
+			$this->post_to_cot_migrator->migrate_orders( $order_ids );
```
Perhaps not a focus of this mvp, but this opens up an opportunity for potential issues if run in parallel.

vedanshujain replied:
We process 20k orders per minute on a good server, so even for 1m orders, we should be able to process within an hour. I am hoping that we won't need it, but I will add if I find that we do.

peterfabian replied:
Yeah, that should be plenty fast. Just in general, this seems like something we should think about more often, to make sure data consistency is paramount.

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,241 @@
+				)
+			);
+
+			$avg_time_per_batch = ( ( $avg_time_per_batch * $batch_count ) + $batch_total_time ) / ( $batch_count + 1 );
```
Avg is not used anywhere. 

You seem to assume you're counting `$batch_count` from 0, but you're counting it from 1, so this will not calculate the avg properly.

Additionally, since you also track `$total_time`, why not just avg = total_time / batch_count (if you calculate avg after updating total_time)?

vedanshujain replied:
Removed `avg`, as you said, we can do `total_time / batch_count` if we need to. See https://github.com/woocommerce/woocommerce/pull/32938/commits/124fdd37f35389dd7ac8ba92255bb159dbc26b9b

peterfabian replied:
👍

peterfabian commented on plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php:
```diff
@@ -5,6 +5,8 @@
 namespace Automattic\WooCommerce\Internal\DataStores\Orders;
 
+use Automattic\Jetpack\Constants;
+use Automattic\WooCommerce\DataBase\Migrations\CustomOrderTable\CLIRunner;
```
Are these necessary here?

vedanshujain replied:
Good catch, removed in https://github.com/woocommerce/woocommerce/pull/32938/commits/80f1a2e815ee8676957fe59ca7cff32da8dffbe8

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,241 @@
+/**
+ * Credits https://github.com/liquidweb/woocommerce-custom-orders-table/blob/develop/includes/class-woocommerce-custom-orders-table-cli.php.
+ *
+ * CLI tool for migrating order data to/from custom table.
```
Might be good to swap those lines as the first on gets displayed with the `wp help wc` command:
![Screenshot 2022-05-10 at 13 28 15](https://user-images.githubusercontent.com/2207451/167618438-066e66d8-c23f-42f5-90aa-61fc5222f2a7.png)

vedanshujain replied:
Done in https://github.com/woocommerce/woocommerce/pull/32938/commits/124fdd37f35389dd7ac8ba92255bb159dbc26b9b

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,241 @@
+	 *
+	 * @internal
+	 */
+	final public function init( CustomOrdersTableController $controller, DataSynchronizer $synchronizer, PostsToOrdersMigrationController $posts_to_orders_migration_controller ) {
```
Since this is public, it gets picked up as a command. Not a huge issue, but would be nice if we could get rid of it at some point.

![Screenshot 2022-05-10 at 13 31 07](https://user-images.githubusercontent.com/2207451/167618881-77a26e1c-6ff2-464d-9191-9622b22eded0.png)

vedanshujain replied:
I have changed how we register the methods, so they shouldn't show up anymore.

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,241 @@
+	private function is_enabled( $log = true ) {
+		if ( ! $this->controller->is_feature_visible() ) {
+			if ( $log ) {
+				WP_CLI::log( __( 'Custom order table usage is not enabled.', 'woocommerce' ) );
```
Would it make sense to point users towards how to enable it?

vedanshujain replied:
agreed, added an intermediate link in https://github.com/woocommerce/woocommerce/pull/32938/commits/ba99395291351a82785309c16f1efef3390cda00

peterfabian replied:
Thanks!

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,241 @@
+	 *
+	 * @return int The number of orders to be migrated.*
+	 */
+	public function count( $log = true ) {
```
For some reason, I'm getting an empty array as `$log` here when I call `wp wc cot count`.  
Thus, it doesn't print anything and the command itself returns 0 (as in successfully completed), so it needs tweaking to be useful (the count this function returns is correct, but there's no way to see it at the moment).

```
% wp wc cot count
% echo $?
0
```

vedanshujain replied:
Fixed in https://github.com/woocommerce/woocommerce/pull/32938/commits/124fdd37f35389dd7ac8ba92255bb159dbc26b9b

peterfabian commented on plugins/woocommerce/includes/class-wc-cli.php:
```diff
@@ -29,6 +29,8 @@ private function includes() {
 		require_once dirname( __FILE__ ) . '/cli/class-wc-cli-update-command.php';
 		require_once dirname( __FILE__ ) . '/cli/class-wc-cli-tracker-command.php';
+		$cli_runner = wc_get_container()->get( \Automattic\WooCommerce\DataBase\Migrations\CustomOrderTable\CLIRunner::class );
+		WP_CLI::add_command( 'wc cot', $cli_runner );
```
This goes slightly against the pattern we have here, where normally the COT cli runner would define a `register_commands` method and run the `WP_CLI::add_command` over there, then hook it down in the `hooks` method to `after_wp_load` instead of running on instantiation. Do you think it would be worth changing it or is it just unnecessary abstraction overhead?

vedanshujain replied:
Thanks for pointing out, changed how we register these method in https://github.com/woocommerce/woocommerce/pull/32938/commits/124fdd37f35389dd7ac8ba92255bb159dbc26b9b

peterfabian reviewed (changes requested):
As an mvp it works. I added a couple of notes if we wanted to improve some aspects. I think the only really important one is to make `count` work as a standalone command, the rest is nice to have.

vedanshujain committed:
Add verification support to custom table migrators.

vedanshujain committed:
Add function to verify migrate orders for COT migration.

vedanshujain committed:
Use `+` to preserve order ID index.

vedanshujain committed:
More defensive checks for date_epoch data type.

vedanshujain committed:
Add verify command in CLI to verify migrated orders.

vedanshujain committed:
Addressed feedback comments.

vedanshujain committed:
Removed unused export.

vedanshujain committed:
Add testing instructions link.

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,412 @@
+				'batch-size' => 500,
+			)
+		);
+		$batch_size  = ( (int) $assoc_args['batch-size'] ) === 0 ? 100 : (int) $assoc_args['batch-size'];
```
Should the default value change to 500 over here as well?

vedanshujain replied:
done in https://github.com/woocommerce/woocommerce/pull/32938/commits/75425a8cc878aa1bfe785d6b1cf9321c37a544d0

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,412 @@
+	 * [--batch-size=<batch-size>]
+	 * : The number of orders to process in each batch.
+	 * ---
+	 * default: 100
```
500

vedanshujain replied:
done in https://github.com/woocommerce/woocommerce/pull/32938/commits/75425a8cc878aa1bfe785d6b1cf9321c37a544d0

peterfabian commented:
Side note: to activate the COT:
1. add this snippet e.g. to wc core functions
```
function enable_cot(){
	$order_controller = wc_get_container()
		->get('Automattic\WooCommerce\Internal\DataStores\Orders\CustomOrdersTableController');
	if( isset( $order_controller ) ) {
		$order_controller->show_feature();
	}
}
add_action( 'init', 'enable_cot', 99 );
```
2. to debug WP CLI, run this command in your shell before running `wp` command: `export XDEBUG_CONFIG="idekey=VVVDEBUG remote_connect_back=1"` (via [this](https://gist.github.com/joshlevinson/93253aec2b41749e10de))

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,412 @@
+	}
+
+	/**
+	 * Helper method to get remaining order count
```
The comment should be updated, copy/pasta error.

vedanshujain replied:
🙌 , addressed in https://github.com/woocommerce/woocommerce/pull/32938/commits/75425a8cc878aa1bfe785d6b1cf9321c37a544d0

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/PostsToOrdersMigrationController.php:
```diff
@@ -90,6 +90,20 @@ public function migrate_orders( array $order_post_ids ): void {
+		return $this->order_table_migrator->verify_migrated_data( $order_post_ids ) +
+			$this->billing_address_table_migrator->verify_migrated_data( $order_post_ids ) +
+			$this->shipping_address_table_migrator->verify_migrated_data( $order_post_ids ) +
+			$this->operation_data_table_migrator->verify_migrated_data( $order_post_ids );
```
Is the meta migration left out intentionally or maybe you planned to add it later?

vedanshujain replied:
yeah, I plan to add meta migration verification later on. initially, I left it out since it is mostly a copy-paste, but it would be good to have verification logic. I will add after first round of public testing.

peterfabian replied:
> I will add after first round of public testing.

I actually think a lot of extension authors might want to test the migration to see if their data got migrated correctly, so either we should explicitly and clearly say this doesn't check custom meta data, or, ideally, add this before the public testing.

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,412 @@
+
+			$order_ids        = $wpdb->get_col(
+				$wpdb->prepare(
+					"SELECT ID FROM $wpdb->posts WHERE post_type = 'shop_order' AND ID > %d ORDER BY ID ASC LIMIT %d",
```
Nice to have: Might be good to factor this out so that `$order_ids` uses the same query as `get_verify_order_count`.

vedanshujain replied:
So these queries are different enough not to be refactored. Specifically, the order_count query has a different select clause, while this query also has additional `limit` and `order by` clauses. It seems cleaner to keep them seperate.

peterfabian replied:
You're right, let's keep it as it is.

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/MetaToCustomTableMigrator.php:
```diff
@@ -570,4 +570,217 @@ private function validate_data( $value, string $type ) {
+	 *
+	 * @return array Array of failed IDs if any, along with columns/meta_key names.
+	 */
+	private function verify_core_columns( $row, $failed_ids ) {
```
Maybe this should be called entity columns, as per our nomenclature in the migration?

vedanshujain replied:
agreed, addressed in https://github.com/woocommerce/woocommerce/pull/32938/commits/75425a8cc878aa1bfe785d6b1cf9321c37a544d0

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/MetaToCustomTableMigrator.php:
```diff
@@ -570,4 +570,217 @@ private function validate_data( $value, string $type ) {
+	private function pre_process_row( $row, $schema, $alias, $destination_alias ) {
+		if ( in_array( $schema['type'], array( 'int', 'decimal' ) ) ) {
+			$row[ $alias ]             = (float) $row[ $alias ];
+			$row[ $destination_alias ] = (float) $row[ $destination_alias ];
```
This is dangerous, and perhaps especially because it works most of the time. 

Not only `float` can't represent `int` or `decimal` exactly, but also [comparing them for equality](https://github.com/woocommerce/woocommerce/pull/32938/files#diff-8b20da85be51922f470a3f51ea880a899ad5a03dbfa87f31f3db79d1fc26afbaR698) is asking for trouble. 

> do not compare floating point numbers directly for equality

via https://www.php.net/manual/en/language.types.float.php

vedanshujain replied:
Great catch, thanks, I changed this use to use `wc_format_decimal` instead in https://github.com/woocommerce/woocommerce/pull/32938/commits/75425a8cc878aa1bfe785d6b1cf9321c37a544d0.

peterfabian replied:
Looks good now, thanks!

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/MetaToCustomTableMigrator.php:
```diff
@@ -570,4 +570,217 @@ private function validate_data( $value, string $type ) {
+
+		$select_clause = implode( ', ', array_merge( $source_select_clauses, $meta_select_clauses, $destination_select_clauses ) );
+
+		$meta_join_clause = implode( ' LEFT JOIN ', $meta_join_clauses );
```
Just double checking here: This can potentially create a query with quite a few `JOIN`s, e.g. the address verification joins postmeta 11 times, the operational verification 15 times. Do you think it might be potentially problematic on sites with huge postmeta (e.g. 1M+ rows)? That said, the speed of this is impressive: on my site with 800k postmeta records (around 0.1s) and EXPLAIN looks cool and searches only a few rows, so it seems to be okay.

vedanshujain replied:
Yeah, this makes for a rather large joint. An alternative approach exists, in which we would fetch metadata, core data, and migrated data in three different queries. We use this approach in the migration logic, if there are issues here, we can change this to use that other approach. But let's keep this as it is for now, that way we will also test both of these approaches and see if one is better than the other. wdyt?

Otherwise it should be fast enough since there are where conditions on both ids as well as meta_keys, so MySQL use either of indexes.

peterfabian replied:
Yeah, let's keep it and see what we get back from testing. 👍

peterfabian reviewed (commented):
Overall, this works great and is super fast, really nice job! A couple more questions/notes below...

Nice to have:  Improvement to the final report by adding error count and mismatched values:
```
Error: 12276 orders were verified in 23.342662 seconds. **#number#** errors were found Array
(
    [12140] => Array
        (
            [0] => {
                        "field": _billing_last_name,
                        "post value": "Batman",
                        "cot value": "Robin"
            }

        )

)
```

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,412 @@
+	 *
+	 * @return bool Whether the COT feature is enabled.
+	 */
+	private function is_enabled( $log = true ) {
```
Minor and non-essential, but here and elsewhere in the change, since these are new methods I wonder if we could add return types (at least in simple cases where the return type is not mixed). Ex:

```suggestion
	private function is_enabled( $log = true ): bool {
```

vedanshujain replied:
thanks, added in https://github.com/woocommerce/woocommerce/pull/32938/commits/75425a8cc878aa1bfe785d6b1cf9321c37a544d0

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,412 @@
+	 * Registers commands for CLI.
+	 */
+	public function register_commands() {
+		WP_CLI::add_command( 'wc cot count', array( $this, 'count' ) );
```
I may be overthinking this (and I also realize that if I run `wp help wc cot count` everything will be clarified) ... but at face value this name is just a little confusing: it suggests a count of the number of rows in the `wc_orders` table (as opposed to a count of the number of rows in `wp_posts` *still to be migrated).*

Possible alternatives, if we wished to make a change:

- `migration-progress`
- `count-unmigrated`
- `count --unmigrated`

And/or what if the `cot` sub-command was renamed to `cot-migration` or `order-migration`?

vedanshujain replied:
renamed the `count` method in https://github.com/woocommerce/woocommerce/pull/32938/commits/c36dc7a294dd82a8a5c0046c1bff6312af70409a but have kept the `cot` as same to keep it less verbose. wdyt?

barryhughes replied:
Noice :+1:

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,412 @@
+		if ( isset( $assoc_args['log'] ) && $assoc_args['log'] ) {
+			WP_CLI::log(
+				sprintf(
+				/* Translators: %1$d is the number of orders to be migrated. */
```
Minor formatting thing:

```suggestion
					/* Translators: %1$d is the number of orders to be migrated. */
```

Funnily enough, this also almost always happens to me when adding translator notes. Not sure why, perhaps a PhpStorm glitch.

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,412 @@
+	 * ## EXAMPLES
+	 *
+	 *     wp wc cot count
+	 *
```
```suggestion
	 *
	 * ## OPTIONS
	 *
	 * [--log=<log>]
	 * : If a message describing the number of orders to be migrated should be output.
	 * ---
	 * default: 1
	 * options:
	 *   - 0
	 *   - 1
	 *
```

Should we document how `--log` works (something like the above, specifying 1 and 0 since `--log=false` actually evaluates as log being true) ... or prefer to keep as a private implementation detail?

vedanshujain replied:
This is actually an internal method, as I don't think anyone would invoke this directly but then won't expect for log to be printed.

barryhughes replied:
Gotcha gotcha

barryhughes commented:
@vedanshujain: not *directly* tied to this PR but I'll leave a note here anyway (happy to move it someplace else—also, apologies if you are already aware) 😁 

During testing, I encountered a doing-it-wrong message regarding this code in the meta-to-meta migrator:

https://github.com/woocommerce/woocommerce/blob/ba99395291351a82785309c16f1efef3390cda00/plugins/woocommerce/src/Database/Migrations/CustomOrderTable/MetaToMetaTableMigrator.php#L256-L269

We're using `$wpdb->prepare()` and we supply a variable to replace a placeholder, but the actual placeholder is missing.

barryhughes reviewed (commented):
Left a few notes, but not much to add beyond what Peter already documented :-)

vedanshujain committed:
Increase batch size to 500 for faster processing.

vedanshujain committed:
Applied PR feedback.

vedanshujain committed:
Rename count to count_unmigrated for better readability.

vedanshujain committed:
Correct formatting for comment.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

vedanshujain commented:
thanks, @peterfabian and @barryhughes, for the review. @peterfabian I have modified the output format as you suggest in https://github.com/woocommerce/woocommerce/pull/32938/commits/75425a8cc878aa1bfe785d6b1cf9321c37a544d0.

@barryhughes applied the suggestions, thanks. also, can you try the migration one more time after truncating all the cot-related tables (especially the metadata one)? Maybe the error will present itself again?

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,418 @@
+				/* Translators: %1$d is the number of migrated orders, %2$f is time taken, %3$f is number of errors and$4%s is formatted array of order ids. */
+					_n( '%1$d order was verified, in %2$f seconds. %3$f error was found: %4$s', '%1$d orders were verified in %2$f seconds. %3$f errors were found %4$s', $processed, 'woocommerce' ),
+					$processed,
+					$total_time,
```
It's missing `count( $failed_ids ),` over here.

vedanshujain replied:
done in https://github.com/woocommerce/woocommerce/pull/32938/commits/387fd1e5cde3cf4950d161b6572ef1bf2349390e

peterfabian reviewed (commented):
Looks good, just one oversight in final reporting. Given the intended audience for testing, I think it would be quite useful to add the verification for meta data as well. Otherwise it works great!

vedanshujain committed:
Add support to migrate meta tables.

vedanshujain commented:
@peterfabian  Added meta verification in https://github.com/woocommerce/woocommerce/pull/32938/commits/387fd1e5cde3cf4950d161b6572ef1bf2349390e

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,520 @@
+
+			return WP_CLI::error(
+				sprintf(
+				/* Translators: %1$d is the number of migrated orders, %2$f is time taken, %3$d is number of errors and$4%s is formatted array of order ids. */
```
```suggestion
					/* Translators: %1$d is the number of migrated orders, %2$f is time taken, %3$d is number of errors and$4%s is formatted array of order ids. */
```

(Minor—indentation)

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,520 @@
+			return WP_CLI::error(
+				sprintf(
+				/* Translators: %1$d is the number of migrated orders, %2$f is time taken, %3$d is number of errors and$4%s is formatted array of order ids. */
+					_n( '%1$d order was verified, in %2$f seconds. %3$f error was found: %4$s', '%1$d orders were verified in %2$f seconds. %3$d errors were found %4$s', $processed, 'woocommerce' ),
```
So first, feel free to ignore these points because it's non-essential stuff ... but I was taking a second pass per your request [to replicate this other problem](https://github.com/woocommerce/woocommerce/pull/32938/#issuecomment-1124591410) and noticed:

1. It might be clearer to write `error(s)` instead of `error` or `errors`, since the choice of text used by `_n()` is modified according to the number of _processed_ items, not the number of errors.
2. I happened to trigger a cascade of errors via this command (as output by `%3$d`), so what I effectively saw was the tail end of a large `print_r()` dump. I realize people in this situation could either scroll-up (if the buffer is large enough) or simply re-run the command and pipe it to a file or something ... but it made me wonder if it would be more user friendly to add text *after* the print_r dump. Ex:

`%1$d orders were verified in %2$f seconds. %3$d errors were found %4$s ↑ Please review the above errors.`

(It would just avoid a little puzzlement, especially if the number of lines of scroll-back are limited—because we won't see the first line of text. This could possibly also apply to some other spots where we do something similar, but again isn't essential.)

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,520 @@
+
+			WP_CLI::debug(
+				sprintf(
+				/* Translators: %1$d is the batch number, %2$d is the batch size. */
```
Minor thing (indentation):

```suggestion
					/* Translators: %1$d is the batch number, %2$d is the batch size. */
```

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,520 @@
+
+			WP_CLI::debug(
+				sprintf(
+				// Translators: %1$d is the batch number, %2$f is time taken to process batch.
```
Minor thing (indentation):

```suggestion
					// Translators: %1$d is the batch number, %2$f is time taken to process batch.
```

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,520 @@
+				$excluded_columns
+			)
+		);
+		$data_to_migrate = $wpdb->get_results( $query, ARRAY_A );
```
```suggestion
		$data_to_validate = $wpdb->get_results( $query, ARRAY_A );
```

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,520 @@
+		$data_to_migrate = $wpdb->get_results( $query, ARRAY_A );
+		// phpcs:enable
+
+		$clubbed_data_to_migrate = $this->club_data( $data_to_migrate );
```
```suggestion
		$clubbed_data_to_validate = $this->club_data( $data_to_validate );
```

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,520 @@
+
+		$clubbed_migrated_data = $this->club_data( $migrated_data );
+
+		foreach ( $clubbed_data_to_migrate as $order_id => $meta ) {
```
```suggestion
		foreach ( $clubbed_data_to_validate as $order_id => $meta ) {
```

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,520 @@
+				// Translators: %1$d is the batch number, %2$f is time taken to process batch.
+					__( 'Batch %1$d (%2$d orders) completed in %3$f seconds', 'woocommerce' ),
+					$batch_count,
+					$batch_total_time
```
I think we are are missing the var for `%2$d` ... should this be:

```suggestion
					__( 'Batch %1$d (%2$d orders) completed in %3$f seconds', 'woocommerce' ),
					$batch_count,
					$processed,
					$batch_total_time
```

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,520 @@
+					$failed_ids[ $order_id ][] = array(
+						'order_id'   => $order_id,
+						'meta_key'   => $meta_key,
+						'meta_value' => $diff,
```
```suggestion
						'orig_meta_value' => $values,
						'new_meta_value'  => isset( $clubbed_migrated_data[ $order_id ][ $meta_key ] ) ? $clubbed_migrated_data[ $order_id ][ $meta_key ] : '',
```

peterfabian commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/CLIRunner.php:
```diff
@@ -0,0 +1,520 @@
+	}
+
+	/**
+	 * Helper method to club response from meta queries into order_id > meta_key.
```
Sorry, can't help myself, is this some meaning of _club_ I'm not familiar with? Maybe @barryhughes as a native speaker can advise if this is understandable 😀

barryhughes replied:
Well, not 100% sure ... but I'm assuming [clubbed into shape](https://classroomclipart.com/images/gallery/Animations/History/caveman_with_club_animation_5A.gif), because we're ensuring the array has a certain structure even in the absence of various fields. 

*Could possibly also allude to [a type of chocolate biscuit from the 1980s](https://www.youtube.com/watch?v=8LmZCMmNPkM), but this seems less likely.*

Could be something to rename? Sidebar (may seem unrelated, but it came to mind because what if we want to change nomenclature further down the line, for other methods that are not private): should all of this be in the public namespace vs the internal namespace?

vedanshujain replied:
Yeah, by `club` I mean the biscuit thingy here. As you can see, just like that chocolate biscuit, we have a layer of order_id and then between two order ids, we have data for failed meta values sandwiched between them.

Actually, let me change it to the sandwich for better readability.

vedanshujain replied:
Or maybe I will change it to say `normalize`, will see.

peterfabian reviewed (dismissed):
Pre-approving, just some small suggestions. Works fine as it is, thanks for all the changes.

barryhughes reviewed (commented):
> @barryhughes applied the suggestions, thanks. also, can you try the migration one more time after truncating all the cot-related tables (especially the metadata one)? Maybe the error will present itself again?

So I think we're safe ... mostly an oversight in my testing (I may have gone a bit rogue in how I approached this) :blush:

For the record, though, after truncating the order tables I then ran `wp option set woocommerce_fake_orders_pending_sync_count 500` and I also added a line of code in the migrate command to reduce that value as each batch was processed.

However, on a subsequent run of the migrate command I failed to reset everything. Therefore:

- Calling [`$this->synchronizer->get_ids_of_orders_pending_sync()`](https://github.com/woocommerce/woocommerce/pull/32938/files#diff-13263072ce10ac901c6669d77abeb53267e74303734f1d78a06ea4c9217c929eR192) yielded an empty array
- This was passed along the line to [`MetaToMetaTableMigrator::get_already_migrated_records()`](https://github.com/woocommerce/woocommerce/blob/387fd1e5cde3cf4950d161b6572ef1bf2349390e/plugins/woocommerce/src/Database/Migrations/CustomOrderTable/MetaToMetaTableMigrator.php#L243) 
- Which then produced [an empty `IN ()` clause](https://github.com/woocommerce/woocommerce/blob/387fd1e5cde3cf4950d161b6572ef1bf2349390e/plugins/woocommerce/src/Database/Migrations/CustomOrderTable/MetaToMetaTableMigrator.php#L243) (and consequently a db error)

I don't think this will happen in practice, it was a result of my shimming around the [temporary fake order count](https://github.com/woocommerce/woocommerce/blob/387fd1e5cde3cf4950d161b6572ef1bf2349390e/plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php#L36-L37). That said ... it probably wouldn't harm to add an extra check (check that `$order_ids` is not empty) [before this line](https://github.com/woocommerce/woocommerce/pull/32938/files#diff-13263072ce10ac901c6669d77abeb53267e74303734f1d78a06ea4c9217c929eR193). Not convinced doing so is *vital*, though.

vedanshujain committed:
Address PR feedback - function name for readability, indentation etc.

vedanshujain commented:
Thank you both addressed all remaining feedback in https://github.com/woocommerce/woocommerce/pull/32938/commits/9d7881c5ba38114eb85e96f66abd70eab7d50c42
